Fix entry tier-up SIMD check in BBQ baseline JIT

ddegazio opened the pull request:
#### f6adb76413178eb10820c04b14cffd501a981e88
<pre>
Fix entry tier-up SIMD check in BBQ baseline JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=252590">https://bugs.webkit.org/show_bug.cgi?id=252590</a>
rdar://105689397

Currently, the new WebAssembly BBQ baseline JIT always passes false for the isSIMD
parameter when calling into the OMG tier-up trigger thunk. This patch makes it so
we track whether the function is a SIMD function using notifyFunctionUsesSIMD(),
and pass that into the thunk generator, ensuring we preserve any SIMD registers
used to pass arguments to the function.

Reviewed by Yusuke Suzuki.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitEntryTierUpCheck):
(JSC::Wasm::BBQJIT::notifyFunctionUsesSIMD):

Canonical link: <a href="https://commits.webkit.org/260562@main">https://commits.webkit.org/260562@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/701f658fce2a0f0525d6c22ec93c9093ff90442b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108718 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/17823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/41571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/4 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/19270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9096 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/100956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114486 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/19270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/41571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/100956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/19270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/41571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/100956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/97846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/10621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/41571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/8742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/41571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/106274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/12969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/106274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3972 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5733,13 +5733,13 @@ class BBQJIT {
         MacroAssembler::Label tierUpResume = m_jit.label();
         auto functionIndex = m_functionIndex;
-        addLatePath([tierUp, tierUpResume, functionIndex](BBQJIT&, CCallHelpers& jit) {
+        bool isSIMD = m_isSIMD;
```
We should get this when late path is executed. `notifyFunctionUsesSIMD` is not called yet when `emitEntryTierUpCheck` is called.

ddegazio replied:
It should be called before `addTopLevel`, so it'll be true if any of our arguments or locals are SIMD (which I think is all we need to handle entry tier-up correctly?). But I do agree it's a better, less fragile solution to move the capture (although capturing members is a pain). I'll fix it up.

Constellation replied:
It is possible that body can include SIMD operations and it generates V128 temp (not local).

ddegazio replied:
Should be resolved in the latest push!

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5733,13 +5733,13 @@ class BBQJIT {
             MacroAssembler::Call call = jit.nearCall();
             jit.jump(tierUpResume);
 
-            bool isSIMD = false; // TODO: Support SIMD
```
So, here, let's get `m_isSIMD` from BBQJIT.

Constellation replied:
So,
```
bool isSIMD = generator.m_isSIMD;
```
and capture this `bool` in `jit.addLinkTask`'s lambda.
`generator` should be `BBQJIT&` which is passed to `addLatePath`'s lambda.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5733,15 +5733,14 @@ class BBQJIT {
-            jit.addLinkTask([=] (LinkBuffer& linkBuffer) {
-                MacroAssembler::repatchNearCall(linkBuffer.locationOfNearCall<NoPtrTag>(call), CodeLocationLabel<JITThunkPtrTag>(Thunks::singleton().stub(triggerOMGEntryTierUpThunkGenerator(isSIMD)).code()));
+            jit.addLinkTask([=, this] (LinkBuffer& linkBuffer) {
+                MacroAssembler::repatchNearCall(linkBuffer.locationOfNearCall<NoPtrTag>(call), CodeLocationLabel<JITThunkPtrTag>(Thunks::singleton().stub(triggerOMGEntryTierUpThunkGenerator(this->m_isSIMD)).code()));
```
I think we cannot use `this` in LinkTask since BBQJIT is already destroyed at that time.
So, see https://github.com/WebKit/WebKit/pull/10383/files#r1112305847, let's get isSIMD at that time, and pass it here.

ddegazio replied:
Oh good catch! Sorry, makes sense.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5733,15 +5733,14 @@ class BBQJIT {
         MacroAssembler::Label tierUpResume = m_jit.label();
         auto functionIndex = m_functionIndex;
-        addLatePath([tierUp, tierUpResume, functionIndex](BBQJIT&, CCallHelpers& jit) {
+        addLatePath([tierUp, tierUpResume, functionIndex, this](BBQJIT&, CCallHelpers& jit) {
```
Instead of capturing `this`, use `BBQJIT&`'s parameter.

ddegazio replied:
This is probably best. Added it in the latest push.

Constellation reviewed (approved):
r=me

ddegazio committed:
Fix entry tier-up SIMD check in BBQ baseline JIT
https://bugs.webkit.org/show_bug.cgi?id=252590
rdar://105689397

Currently, the new WebAssembly BBQ baseline JIT always passes false for the isSIMD
parameter when calling into the OMG tier-up trigger thunk. This patch makes it so
we track whether the function is a SIMD function using notifyFunctionUsesSIMD(),
and pass that into the thunk generator, ensuring we preserve any SIMD registers
used to pass arguments to the function.

Reviewed by Yusuke Suzuki.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitEntryTierUpCheck):
(JSC::Wasm::BBQJIT::notifyFunctionUsesSIMD):

Canonical link: https://commits.webkit.org/260562@main
